File: src/main/java/com/onezero/oop/settlement/SettlementStatus.java
package com.onezero.oop.settlement;

import java.util.EnumSet;
import java.util.Set;

public enum SettlementStatus {
    OPENED,
    AGGREGATING,
    VALIDATING,
    CONFIRMED,
    FAILED,
    CORRECTION_REQUESTED,
    CORRECTED;

    public boolean canTransitionTo(SettlementStatus next) {
        return allowedTargets().contains(next);
    }

    private Set<SettlementStatus> allowedTargets() {
        return switch (this) {
            case OPENED -> EnumSet.of(AGGREGATING, FAILED);
            case AGGREGATING -> EnumSet.of(VALIDATING, FAILED);
            case VALIDATING -> EnumSet.of(CONFIRMED, FAILED);
            case CONFIRMED -> EnumSet.of(CORRECTION_REQUESTED);
            case CORRECTION_REQUESTED -> EnumSet.of(CORRECTED, FAILED);
            case CORRECTED, FAILED -> EnumSet.noneOf(SettlementStatus.class);
        };
    }
}

File: src/main/kotlin/com/onezero/oop/settlement/SettlementStatus.kt
package com.onezero.oop.settlement

enum class SettlementStatus {
    OPENED,
    AGGREGATING,
    VALIDATING,
    CONFIRMED,
    FAILED,
    CORRECTION_REQUESTED,
    CORRECTED,
    ;

    fun canTransitionTo(next: SettlementStatus): Boolean = next in allowedTargets()

    private fun allowedTargets(): Set<SettlementStatus> =
        when (this) {
            OPENED -> setOf(AGGREGATING, FAILED)
            AGGREGATING -> setOf(VALIDATING, FAILED)
            VALIDATING -> setOf(CONFIRMED, FAILED)
            CONFIRMED -> setOf(CORRECTION_REQUESTED)
            CORRECTION_REQUESTED -> setOf(CORRECTED, FAILED)
            CORRECTED, FAILED -> emptySet()
        }
}
